pallets/price-fetch/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+		let mut samples_count = Price::from(0);
+		price_points.iter().for_each(|pp| {
+			sum = sum + pp.price;
+			samples_count = samples_count + Price::from(1);
+		});
+
+		let avg_price = sum / samples_count;
+
+		let results = signer.send_signed_transaction(|_account| {
+			// Received price is wrapped into a call to `submit_price` public function of this pallet.
+			// This means that the transaction, when executed, will simply call that function passing
+			// `price` as an argument.
+			Call::submit_new_avg_price(fetcher.symbol.clone(), avg_price)
+		});
+
+		for (acc, res) in &results {
+			match res {
+				Ok(()) => debug::info!("New price submitted by [{:?}]", acc.id),
+				Err(e) => debug::error!("[{:?}] Failed to submit transaction: {:?}", acc.id, e),
+			}
+		}
+
+		Ok(())
+	}
+
+	fn fetch_price_and_submit(fetcher: Fetcher<T::BlockNumber>) -> Result<(), &'static str> {
+		let signer = Signer::<T, T::AuthorityId>::all_accounts();
+		if !signer.can_sign() {
+			return Err("No local accounts available. Consider adding one via `author_insertKey` RPC.");
+		}
+
+		//NOTE: Blocking http request
+		let fetched_price = Self::fetch_price(fetcher.url).map_err(|_| "Failed to fetch data")?;
+
+		let results = signer.send_signed_transaction(|_account| {
+			// Received price is wrapped into a call to `submit_price` public function of this pallet.
+			// This means that the transaction, when executed, will simply call that function passing
+			// `price` as an argument.
+			Call::submit_new_price(fetched_price.clone())
+		});
+
+		for (acc, res) in &results {
+			match res {
+				Ok(()) => debug::info!("New price submitted by [{:?}]", acc.id),
+				Err(e) => debug::error!("[{:?}] Failed to submit transaction: {:?}", acc.id, e),
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Fetch current price from url
+	fn fetch_price(url: Vec<u8>) -> Result<DiaPriceRecord, http::Error> {
+		// deadline to complete the external call.
+		let deadline = sp_io::offchain::timestamp().add(Duration::from_millis(2_000));
+
+		let request = http::Request::get(sp_std::str::from_utf8(&url).unwrap());
+		let pending = request.deadline(deadline).send().map_err(|_| http::Error::IoError)?;
+
+		let response = pending.try_wait(deadline).map_err(|_| http::Error::DeadlineReached)??;
+
+		if response.code != 200 {
+			debug::warn!("Unexpected status code: {}", response.code);
+			return Err(http::Error::Unknown);
+		}
+
+		let body = response.body().collect::<Vec<u8>>();
+		let body_str = sp_std::str::from_utf8(&body).map_err(|_| {
+			debug::warn!("No UTF8 body");
+			http::Error::Unknown
+		})?;
+
+		let price = match Self::parse_dia_res(body_str) {
+			Some(price) => Ok(price),
+			None => {
+				debug::warn!("Unable to parse response: {:?}", body_str);
+				Err(http::Error::Unknown)
+			}
+		}?;
+
+		Ok(price)
+	}
+
+	/// Parse json response body received from dia request
+	///
+	/// Returns `None` when parsing failed or `Some(DiaPriceRecord)` when parsing is successful.
```

Comment:
na, it panics xD

Reply:
yes, you're right. This comment is how function should work at the end :smile:  I'll fix it

Reply:
fixed https://github.com/galacticcouncil/hack.HydraDX-node/blob/83b6e02051cc2f3d19b1b351f92174b7d4f83938/pallets/price-fetch/src/lib.rs#L388